### ISSUE TITLE
Allow a custom (extended) WebDriver class

### ISSUE THREAD

**Issue by zoechi** [google/webdriver.dart#96]
I haven't found a way to create a custom "WebDriver" class or extend `WebDriver`.

I would like to have an extended `WebDriver` class with additional methods (for example using common JS snippets with `execute()`) but I have to copy a lot of code to make it work. 

The main problems so far are `_CommandProcessor` being private and `WebElement`s private constructor. The next problem is probably `_WebDriverBase`

**Comment by zoechi** [comment]
I got it working by embedding the classes of this package within my own and forward all calls. This is quite verbose and cumbersome. It would be nice if there were an easier way.

**Comment by DrMarcII** [comment]
Your approach with delegation was actually what I was going to suggest (I just got distracted with life and hadn't had a chance to respond). In general, I prefer composition over inheritance, so I designed these classes with that in mind.

So why did you find the delegation approach verbose? You can trivially forward all calls using noSuchMethod, and then only have to write code for any new methods you added or existing methods that you want to modify.

Where I can see problems now are with _post, _get, and _delete being private on _WebDriverBase (if you are also creating delegating wrappers for non-WebDriver classes such as WebElement), and I have no problem with making these public if needed. 

Additionally, the .driver field of any object that extends _WebDriverBase will point to the original WebDriver instance, not to your wrapper. If this causes you problems, I am sure that we could come up with a way to work around it.

**Comment by zoechi** [comment]
I guess you're right with composition being the better approach anyway. 
I wrapped some of the classes already and didn't run into any further issues so far.
If something comes up I'll create a new issue.
I banned `noSuchMethod` from my memory because of the issues it causes on client side (code bloat), but I think it's time to reconsider, I have no plans running this code in the browser anyway. 
Thanks for your time!
